src/iterators/mod.rs

```diff
@@ -1182,9 +1195,13 @@ impl<'a, A, D: Dimension> NdProducer for AxisIterMut<'a, A, D> {
 /// See [`.axis_chunks_iter()`](../struct.ArrayBase.html#method.axis_chunks_iter) for more information.
 pub struct AxisChunksIter<'a, A, D> {
     iter: AxisIterCore<A, D>,
-    n_whole_chunks: usize,
-    /// Dimension of the last (and possibly uneven) chunk
-    last_dim: D,
+    /// Index of the partial chunk (the chunk smaller than the specified chunk
+    /// size due to the axis length not being evenly divisible). If the axis
+    /// length is evenly divisible by the chunk size, this index is larger than
+    /// the maximum valid index.
+    partial_chunk_index: usize,
```

Comment:
Would it be beneficial to rephrase this as an `Option`, to make it clearer that we might (or might not) have a partial chunk? Something along the lines of:
```rust
pub struct AxisChunksIter<'a, A, D> {
   iter: AxisIterCore<A, D>,
   partial_chunk: Option<PartialChunk>,
   life: PhantomData<&'a A>
}

struct PartialChunk {
   partial_chunk_index: usize,
   partial_chunk_dim: D
}
```

Reply:
I don't think it makes sense to use *both* the `Option` variant and the value of `partial_chunk_index` to represent whether or not there's a partial chunk. (The biggest reason is that I prefer data structures where there's a single source of truth, rather than having to keep multiple things in sync. There might also be a small performance cost to accessing `partial_chunk_index` through the `Option` (since accessing it requires checking whether the `Option` is the `Some` variant), but we'd need to test to determine if that would really be noticeable.) IMO, putting the fields in an `Option` would be additional complication over the current approach without much benefit.

It would be reasonable to eliminate `partial_chunk_index` and just use the `Option` variant to represent the presence of a partial chunk, like this:

```rust
pub struct AxisChunksIter<'a, A, D> {
   iter: AxisIterCore<A, D>,
   partial_chunk: Option<D>,
   life: PhantomData<&'a A>
}
```

or to always store the shape of the last chunk (regardless of whether or not it's a partial chunk):

```rust
pub struct AxisChunksIter<'a, A, D> {
   iter: AxisIterCore<A, D>,
   last_chunk_dim: D,
   life: PhantomData<&'a A>
}
```

These approaches have two disadvantages since they rely on checking whether the iterator is at its end to handle the partial chunk instead of checking whether the current index is equal to `partial_chunk_index`:

1. `.split_at()` needs to check whether or not the partial chunk is in the left piece and determine `partial_chunk` or `last_chunk_dim` of the left piece accordingly. (The partial chunk is in the left piece when `index == self.iter.len()`.)

2. `.next_back()` needs to set `partial_chunk` to `None` or `last_chunk_dim` to `self.iter.inner_dim` each time it's called.

So, I'd rather keep the current approach and add more comments if necessary to make it clear.